kernel/cbytegen.ml

```diff
@@ -873,38 +736,63 @@ and compile_universe reloc uni sz cont =
+    let u = Univ.Instance.to_array u in
+    let lu = Array.length u in
+    let all =
+      Array.init (lu + Array.length args)
+        (fun i -> if i < lu then ArgUniv u.(i) else ArgLambda args.(i-lu))
+    in
+    comp_app (fun _ _ _ cont -> Kgetglobal kn :: cont)
+      compile_arg reloc () all sz cont
+
+(*template for n-ary operation, invariant: n>=1,
+  the operations does the following :
+  1/ checks if all the arguments are constants (i.e. non-block values)
+  2/ if they are, uses the "op" instruction to execute
+  3/ if at least one is not, branches to the normal behavior:
+      Kgetglobal (get_alias !global_env kn) *)
+and op_compilation env n op =
+  let code_construct reloc kn sz cont =
+     let f_cont =
+         let else_lbl = Label.create () in
+         Kareconst(n, else_lbl):: Kacc 0:: Kpop 1::
+          op:: Kreturn 0:: Klabel else_lbl::
+         (* works as comp_app with nargs = n and tailcall cont [Kreturn 0]*)
+          compile_get_global reloc kn sz (
+            Kappterm(n, n):: []) (* = discard_dead_code [Kreturn 0] *)
+     in
+     let lbl = Label.create () in
+     fun_code := [Ksequence (add_grab n lbl f_cont, !fun_code)];
+     Kclosure(lbl, 0)::cont
+  in
+  fun kn reloc args sz cont ->
   let nargs = Array.length args in
-  (* spiwack: checks if there is a specific way to compile the constant
-              if there is not, Not_found is raised, and the function
-              falls back on its normal behavior *)
-  try
-    Retroknowledge.get_vm_compiling_info (!global_env).retroknowledge
-                  (mkConstU (kn,u)) reloc args sz cont
-  with Not_found ->
-    if Int.equal nargs 0 then
-      compile_get_global reloc (kn,u) sz cont
-    else
-      if Univ.Instance.is_empty u then
-        (* normal compilation *)
-        comp_app (fun _ _ sz cont ->
-            compile_get_global reloc (kn,u) sz cont)
-          compile_constr reloc () args sz cont
-      else
-        let compile_arg reloc constr_or_uni sz cont =
-          match constr_or_uni with
-          | ArgConstr cst -> compile_constr reloc cst sz cont
-          | ArgUniv uni -> compile_universe reloc uni sz cont
-        in
-	let u = Univ.Instance.to_array u in
-	let lu = Array.length u in
-	let all =
-	  Array.init (lu + Array.length args) 
-	    (fun i -> if i < lu then ArgUniv u.(i) else ArgConstr args.(i-lu))
-	in
-        comp_app (fun _ _ _ cont -> Kgetglobal kn :: cont)
-          compile_arg reloc () all sz cont
+  if Int.equal nargs n then (*if it is a fully applied addition*)
+    let (escape, labeled_cont) = make_branch cont in
+    let else_lbl = Label.create () in
+      comp_args (compile_lam env) reloc args sz
+        (Kisconst else_lbl::(make_areconst (n-1) else_lbl
+           (*Kaddint31::escape::Klabel else_lbl::Kpush::*)
+           (op::escape::Klabel else_lbl::Kpush::
+           (* works as comp_app with nargs = n and non-tailcall cont*)
+           compile_get_global reloc kn sz (Kapply n::labeled_cont))))
```

Comment:
We need `assert (n < 4)` here. Otherwise the generated bytecode is plain wrong.

Reply:
Also `sz` should be replaced by `(sz + n)` in the call to `compile_get_global`. Even with the 15 extra stack words, we are getting a bit reckless here.

Reply:
Good points. Note that these problems were already there before this PR. Will fix.